Added the initial code for gdeploy wrapper

nnDarshan opened the pull request:
This patch adds a wrapper for gdeploy, which can be used to
generate gdeploy config files and further execute them to
get the desired gluster provisoning tasks done. Also
this patch adds actions like :
* setup gluster node
* gluster cluster creation
* gluster brick provison
* gluster volume creation
* volume option configuration
* snapshot configuration
* quota configuration
* gluster volume expansion

NOTE: this package is independent of any other tendrl
packages, this package can be independently installed
and above mentioned actions can be run by loading the
suitable modules.

Example:

```
from python_gdeploy.actions import setup_gluster_node
out, err, rc = setup_gluster_node.setup_gluster_node(<host_list>)
```
This will install all glusterfs related packages on the
hosts provided in the host list, start the glusterd service
and configure the firewalld. Similarly other actions
can also be performed

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

nnDarshan committed:
Added the initial code for gdeploy wrapper

This patch adds a wrapper for gdeploy, which can be used to
generate gdeploy config files and further execute them to
get the desired gluster provisoning tasks done. Also
this patch adds actions like :
* setup gluster node
* create cluster
* gluster brick provison
* gluster volume creation
* volume option configuration
* snapshot configuration
* quota configuration
* gluster volume expansion

NOTE: this package is independent of any other tendrl
packages, this package can be independently installed
and above mentioned actions can be run by loading the
suitable modules.

Example:

from python_gdeploy.actions import \
    setup_gluster_node
out,err,rc=setup_gluster_node. \
    setup_gluster_node(<host_list>)

This will install all glusterfs related packages on the
hosts provided in the host list, start the glusterd service
and configure the firewalld. Similarly other actions
can also be performed

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

nnDarshan commented:
@brainfunked @r0h4n @nthomas-redhat @shtripat Please review

shtripat commented on tendrl/python_gdeploy/actions/create_gluster_volume.py:
```diff
@@ -0,0 +1,43 @@
+
+    config_str = cook_gdeploy_config(recipe)
+
+    out, err, rc = invoke_gdeploy(config_str)
```
Dont we return these values?

nnDarshan replied:
Yes, it has to be returned. Had missed that, thanks for pointing

shtripat commented on tendrl/python_gdeploy/actions/gluster_package_installation.py:
```diff
@@ -0,0 +1,38 @@
+    "glusterfs-fuse"
+]
+
+GLUSTERFS_REPO = "https://buildlogs.centos.org/centos/7/"\
```
Would it be same in case of downstream as well? dont think so. If yes it should be configurable ..

nnDarshan replied:
good catch. Have maintained a conf file for python-gdeploy in latest PR

shtripat reviewed (commented):
LGTM

r0h4n commented on python_gdeploy/wrapper/gdeploy_wrapper.py:
```diff
@@ -0,0 +1,43 @@
+    return "[" + section + "]\n"
+
+
+def cook_gdeploy_config(recipe):
```
What is the schema of the config? INI style? Please use appropriate config parsers (yaml or ini or json)

nnDarshan replied:
It resembles INI style. But its very different from it. for example; it has few sections where there will be only keys (no key=value pair as in case of INI), sections where value is a list and not a single value etc. Similarly there are quite a few variations from INI.  So could not use the INI parser

r0h4n commented on tox.ini:
```diff
@@ -0,0 +1,37 @@
+[testenv:cover]
+passenv = TRAVIS TRAVIS_JOB_ID TRAVIS_BRANCH
+commands =
+    py.test --cov-report= --cov=python_gdeploy python_gdeploy/tests
```
Check node-agent develop branch for new code coverage tool changes to tox.ini and requirements etc

https://github.com/Tendrl/node-agent/blob/develop/tox.ini#L30

nnDarshan replied:
Done

nnDarshan committed:
Adds unit tests and fixes code coverage conf

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

nnDarshan committed:
Adds more unit tests.

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

nnDarshan committed:
Fixed pep8 issues

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

nnDarshan committed:
Setting ignore errors flag to false

This patch sets ignore_<action>_errors flag to no,
so that the errors while performing actions are not
silently ignored.

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

nnDarshan commented:
@r0h4n @nthomas-redhat @shtripat Please review

nnDarshan committed:
Split some actions to make it more atomic

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

r0h4n commented on etc/python-gdeploy.conf:
```diff
@@ -0,0 +1,2 @@
+[python-gdeploy]
+glusterfs_repo = https://buildlogs.centos.org/centos/7/storage/x86_64/gluster-3.9/
```
is this a sample file? this shouldnt be hardcoded if not a sample file

nnDarshan replied:
okay, will make this a sample file and ask users to copy it and modify accordingly if needed

r0h4n commented on python_gdeploy/actions/configure_gluster_firewall.py:
```diff
@@ -0,0 +1,34 @@
+from python_gdeploy.wrapper.gdeploy_wrapper import invoke_gdeploy
+
+
+def configure_gluster_firewall(host_list):
```
please put a comment with sample value of "host_list"

nnDarshan replied:
will do that

r0h4n commented on python_gdeploy/actions/create_gluster_volume.py:
```diff
@@ -0,0 +1,47 @@
+from python_gdeploy.wrapper.gdeploy_wrapper import invoke_gdeploy
+
+
+def create_volume(volume_name, brick_details, transport=[],
```
please add sample of "brick_details"

nnDarshan replied:
its added as a comment below

r0h4n commented on python_gdeploy/actions/create_gluster_volume.py:
```diff
@@ -0,0 +1,47 @@
+            "create",
+            brick_dirs=brick_list,
+            transport=transport,
+            replica_count=replica_count,
```
no validations for replica_count or other attributes?

nnDarshan replied:
will add that

r0h4n commented on python_gdeploy/actions/install_gluster.py:
```diff
@@ -0,0 +1,46 @@
+
+def get_glusterfs_repo():
+    config = ConfigParser.SafeConfigParser()
+    config.read('/etc/python-gdeploy/python-gdeploy.conf')
```
who is going to generate this file?

nnDarshan replied:
As mentioned above in dev setup user will have to, copy the sample conf file. In rpm based instance, during installation this will be copied.
In any case this conf file will not be of any use if we invoke python-gdeploy from tendrl as we will pass details like repo, rpm names to be installed from tendrl.

r0h4n commented on python_gdeploy/actions/install_gluster.py:
```diff
@@ -0,0 +1,46 @@
+
+def install_gluster_packages(host_list,
+                             glusterfs_packages=None,
+                             glusterfs_repo=None,
```
who calls install_gluster_packages and provides the package list?

nnDarshan replied:
the provisioning wrapper calls this method and it has to pass glusterfs repos and glusterfs package list. I feel we need to maintain these details in master definition file, under the import cluster flow. 

Correct me if my understanding is wrong

r0h4n replied:
Yes, please make that change to tendrl.objects.definitions.master.yml in commons

r0h4n commented on python_gdeploy/wrapper/gdeploy_wrapper.py:
```diff
@@ -0,0 +1,48 @@
+    return config_str
+
+
+def invoke_gdeploy(config):
```
Is this function thread/greenthread safe? What happens if we invoke "gdeploy -c conf_file" from multiple threads at the same time?

nnDarshan replied:
Yes, this is thread safe as: we are having a uuid appended at the end of each conf file, each call to this function will have its own conf file. Also in the backend "gdeploy -c" can handle multiple simultaneous calls without any issue.

nnDarshan committed:
Patch handles proper brick placement for Distributed
replicated volume

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

nnDarshan committed:
Patch adds more gluster actions

This patch adds more gluster actions like:
volume start/stop/delete/rebalance/shrink/expand
and peer addtion and deletion

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

nnDarshan committed:
Modified brick creation to take lvm related details

This patch modifies brick creation action to accept lvm
related details like lv name, pv name, vg name, pool name

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

cybernth commented on python_gdeploy/wrapper/gdeploy_wrapper.py:
```diff
@@ -0,0 +1,48 @@
+
+def invoke_gdeploy(config):
+    conf_file = GDEPLOY_CONFIG_PATH + str(uuid.uuid4()) + ".conf"
+    with open(conf_file, 'w') as f:
```
Handle exceptions during file write and subprocess.popen. os.remove(conf_file) should be in a finally block

cybernth commented on requirements.txt:
```diff
@@ -0,0 +1 @@
+setuptools
```
Now this part of setup.py. Something like 
install_requires=[
        
        ],
Please refer https://github.com/Tendrl/node-agent/blob/develop/setup.py#L111
patch : https://github.com/Tendrl/node-agent/commit/e5d98d1fe9e2d1d676bfd6bbe780df177d649609

nnDarshan replied:
Done

cybernth commented on setup.py:
```diff
@@ -0,0 +1,26 @@
+module_file = read_module_contents()
+metadata = dict(re.findall("__([a-z]+)__\s*=\s*'([^']+)'", module_file))
+version = metadata['version']
+
```
Please add the support for version management as well
refer https://github.com/Tendrl/node-agent/blob/develop/setup.py#L27

nnDarshan replied:
Done

r0h4n commented on .travis.yml:
```diff
@@ -0,0 +1,15 @@
+---
```
Change as per https://github.com/Tendrl/commons/blob/develop/.travis.yml

nnDarshan replied:
Done

r0h4n commented on etc/python-gdeploy.conf.sample:
```diff
@@ -0,0 +1,2 @@
+[python-gdeploy]
```
put this file in etc/python-gdeploy/python-gdeploy.conf.sample

nnDarshan replied:
Done

r0h4n commented on python_gdeploy/wrapper/gdeploy_wrapper.py:
```diff
@@ -0,0 +1,48 @@
+import subprocess
+import uuid
+
+GDEPLOY_CONFIG_PATH = "/var/run/python_gdeploy/config_"
```
is this path being maintained in this project's spec file?

nnDarshan replied:
@TimothyAsir is working on building the spec file. I think he is taking care of this

r0h4n commented on setup.py:
```diff
@@ -0,0 +1,26 @@
+import re
```
Change as per https://github.com/Tendrl/commons/blob/develop/setup.py

nnDarshan replied:
Done

r0h4n commented on tox.ini:
```diff
@@ -0,0 +1,40 @@
+[tox]
```
Change as per https://github.com/Tendrl/commons/blob/develop/tox.ini

nnDarshan replied:
Done

cybernth commented on python_gdeploy/actions/create_gluster_volume.py:
```diff
@@ -0,0 +1,129 @@
+    if transport:
+        arg_dict.update({"transport": transport})
+    if replica_count:
+        arg_dict.update({"replica_count": replica_count})
```
Do you need to pass replica 'yes' here? Or by default its set?

nnDarshan replied:
If count is mentioned..its treated as a replicated volume. No need to pass replica=yes explicitly

cybernth commented on python_gdeploy/actions/create_gluster_volume.py:
```diff
@@ -0,0 +1,129 @@
+        arg_dict.update({"disperse_count": disperse_count,
+                         "redundancy_count": redundancy_count,
+                         "disperse": "yes"})
+
```
what about volume options like client access etc?
gdeploy provides volume create with vol options in a single go right?

nnDarshan replied:
Yes, but in backend even gdeploy creates a volume first and then sets the options after volume is created. Since volume create is pretty complex with many conditions as of now...I feel its better to have them as separate actions.

nnDarshan committed:
Modified setup.py, tox.ini and .travis.yml to be in sync
with other projects

tendrl-bug-id: Tendrl/python-gdeploy#1
Signed-off-by: nnDarshan <darshan.n.2024@gmail.com>

cybernth commented on python_gdeploy/actions/volume_quota_config.py:
```diff
@@ -0,0 +1,47 @@
+from python_gdeploy.wrapper.gdeploy_wrapper import invoke_gdeploy
+
+
+def volume_quota_config(volume_name, hostname, action, dir_details={}):
```
what are the supported actions?

nnDarshan replied:
limit-usage and limit-objects... Will add these as comments

cybernth commented on python_gdeploy/actions/volume_quota_config.py:
```diff
@@ -0,0 +1,47 @@
+        for directory in dir_details:
+            dir_list.append(directory.keys()[0])
+            size_list.append(directory.values()[0])
+        recipe.append(
```
add/remove directory quota supported here?

nnDarshan replied:
this method currently supports only 2 actions namely limit-usage and limit-objects. Others have to be added as and when need arises. Have added a TODO for this

cybernth commented on python_gdeploy/actions/volume_quota_config.py:
```diff
@@ -0,0 +1,47 @@
+            )
+        )
+    else:
+        recipe.append(
```
I think enable, disbale quota's go in here.
what about soft-limit, timout, alerts etc?

nnDarshan replied:
this method currently supports only 2 actions namely limit-usage and limit-objects. Others have to be added as and when need arises. Have added a TODO for this

cybernth commented on python_gdeploy/actions/snapshot_config.py:
```diff
@@ -0,0 +1,21 @@
+from python_gdeploy.wrapper.gdeploy_wrapper import invoke_gdeploy
+
+
+def volume_snapshot_config(volume_name, hostname, action, snapname=""):
```
What actions supported here?

nnDarshan replied:
currently only snapshot create,delete and clone are supported. Have added a TODO to support other features
